ctru-rs/src/applets/swkbd.rs

```diff
@@ -675,6 +640,292 @@ impl SoftwareKeyboard {
+                shared_mem_size as _,
+                MEMPERM_READ | MEMPERM_WRITE,
+                MEMPERM_READ | MEMPERM_WRITE,
+            )
+        };
+
+        if R_FAILED(res) {
+            unsafe {
+                libc::free(swkbd_shared_mem_ptr);
+                swkbd.result = SWKBD_OUTOFMEM;
+                return SWKBD_BUTTON_NONE;
+            }
+        }
+
+        // Copy stuff to shared mem
+        if !extra.initial_text.is_null() {
+            swkbd.initial_text_offset = 0;
+
+            unsafe {
+                let utf16_iter =
+                    str::from_utf8_unchecked(CStr::from_ptr(extra.initial_text).to_bytes())
+                        .encode_utf16()
+                        .take(swkbd.max_text_len as _)
+                        .chain(once(0));
+
+                let mut initial_text_cursor = swkbd_shared_mem_ptr.cast();
+
+                for code_point in utf16_iter {
+                    *initial_text_cursor = code_point;
+                    initial_text_cursor = initial_text_cursor.add(1);
+                }
+            }
+        }
+
+        if !extra.dict.is_null() {
+            swkbd.dict_offset = dict_off as _;
+            unsafe {
+                std::ptr::copy_nonoverlapping(
+                    extra.dict,
+                    swkbd_shared_mem_ptr.add(dict_off).cast(),
+                    swkbd.dict_word_count as _,
+                )
+            };
+        }
+
+        if swkbd.initial_status_offset >= 0 {
+            swkbd.initial_status_offset = status_off as _;
+            unsafe {
+                std::ptr::copy_nonoverlapping(
+                    extra.status_data,
+                    swkbd_shared_mem_ptr.add(status_off).cast(),
+                    1,
+                )
+            };
+        }
+
+        if swkbd.initial_learning_offset >= 0 {
+            swkbd.initial_learning_offset = learning_off as _;
+            unsafe {
+                std::ptr::copy_nonoverlapping(
+                    extra.learning_data,
+                    swkbd_shared_mem_ptr.add(learning_off).cast(),
+                    1,
+                )
+            };
+        }
+
+        if extra.callback.is_some() {
+            swkbd.filter_flags |= SWKBD_FILTER_CALLBACK;
+        } else {
+            swkbd.filter_flags &= !SWKBD_FILTER_CALLBACK;
+        }
+
+        // Launch swkbd
+        unsafe {
+            swkbd.__bindgen_anon_1.reserved.fill(0);
+
+            if extra.callback.is_some() {
+                self.callback_data.extra = std::ptr::addr_of_mut!(extra);
+                self.callback_data.swkbd_shared_mem_ptr = swkbd_shared_mem_ptr;
+
+                aptSetMessageCallback(
+                    Some(Self::swkbd_message_callback),
+                    std::ptr::addr_of_mut!(self.callback_data).cast(),
+                );
+            }
+
+            aptLaunchLibraryApplet(
+                APPID_SOFTWARE_KEYBOARD,
+                swkbd as *mut _ as *mut _,
```

Comment:
can at least one of these be a `cast()`/`cast_mut()`? Presumable you can `as *mut T` and then cast to `c_void`?